src/coreclr/jit/optimizebools.cpp

```diff
@@ -2022,8 +1663,134 @@ PhaseStatus Compiler::optOptimizeBools()
         }
     } while (change);
 
-    JITDUMP("\noptimized %u BBJ_COND cases, %u BBJ_RETURN cases in %u passes\n", numCond, numReturn, numPasses);
+    JITDUMP("\noptimized %u BBJ_COND cases in %u passes\n", numCond, numPasses);
 
-    const bool modified = stress || ((numCond + numReturn) > 0);
+    const bool modified = stress || (numCond > 0);
     return modified ? PhaseStatus::MODIFIED_EVERYTHING : PhaseStatus::MODIFIED_NOTHING;
 }
+
+//-------------------------------------------------------------
+// fgFoldCondToReturnBlock: Folds BBJ_COND <relop> into BBJ_RETURN <relop>
+//   This operation is the opposite of what fgDedupReturnComparison does.
+//   We don't fold such conditionals if both return blocks have multiple predecessors.
+//
+// Arguments:
+//    block - the BBJ_COND block to convert into BBJ_RETURN <relop>
+//
+// Returns:
+//    true if the block was converted into BBJ_RETURN <relop>
+//
+bool Compiler::fgFoldCondToReturnBlock(BasicBlock* block)
+{
+    bool modified = false;
+
+    assert(block->KindIs(BBJ_COND));
+
+#ifdef JIT32_GCENCODER
+    // JIT32_GCENCODER has a hard limit on the number of epilogues.
+    return modified;
+#endif
+
+    // Early out if the current method is not returning a boolean.
+    if ((info.compRetType != TYP_UBYTE) || (genReturnBB != nullptr))
```

Comment:
Seems like we could proceed and then bail out later, if one of the successors is `genReturnBB`.

Reply:
@AndyAyersMS from my understanding, if `genReturnBB` is not null, we're guaranteed that there is nothing we can do - there will be no RETURN true/RETURN false blocks out of BBJ_COND, so this check is purely to improve TP and give up early

Reply:
Although, I misunderstood it. From my understanding, when we run out of some arbitrary limit of RETURNs in the merge-return phase, we merge them using this `genReturnBb` so you're right, we should not give up here

Reply:
Addressed via https://github.com/dotnet/runtime/pull/107499/commits/ce10e5c8665147cf8ff11900e04402eeaff23364